### ISSUE TITLE
Too Many SOQLs Error 

### ISSUE THREAD

**Issue by jimrae** [ForceDotComLabs/QuotaEdit#21]
I have a large org with a complex hierarchy, when I try to apply my changes to a new period, I get a too many soqls error. I am assuming that this is because of the recursive call to the getAllSubRoleIds method.
```Too many SOQL queries: 101
Error is in expression '{!applyToPeriod}' in component <apex:commandButton> in page quotaedit: Class.RoleUtils.getRoleSubordinateUsers: line 12, column 1
Class.QuotaController.getQuotasForPeriod: line 159, column 1
Class.QuotaController.applyToPeriod: line 382, column 1```

Any suggestions on how to address this?

**Comment by alonkama** [comment]
Wow, I guess I didn't anticipate this. I suppose we could limit how deep the recursion goes, but that will result in your seeing less users than are available. To be clear, the "too many SOQL queries" error comes from Salesforce: we can't run so many in one operation.

getRoleSubordinateUsers() is bypassed in https://github.com/ForceDotComLabs/QuotaEdit/blob/7fa5bc8163a908e26ed008ecddce556b6e1671fb/src/classes/QuotaController.cls#L142 if the user has ViewAllForecasts user permission. Is that possible for you?

An alternative is to delegate the setting of quotas to people further below you in the hierarchy, where conceivably they won't hit the limit.

**Comment by alonkama** [comment]
I read about a different way I can go about this: querying all Roles, then putting together a tree using the ParentRoleId field. If the workaround I gave above isn't viable, let me know and I'll try that change.

**Comment by jimrae** [comment]
Actually, I figured out a different way to address the issue.  I had modified the base code to provide the ability to apply the submitted quotas across a range of periods (instead of just one).  This was causing the getRoleSubordiateUsers to fire many times. I modified that class to use a "lazy loading" style, where if the list was not null, bypass the recursive query set.  If you are interested, I could make a pull request and submit my changes to your repo.
